includes/AMP/Optimization.php

```diff
@@ -0,0 +1,145 @@
+use Google\Web_Stories_Dependencies\AmpProject\Optimizer\Transformer\TransformedIdentifier;
+use Google\Web_Stories_Dependencies\AmpProject\RemoteRequest\FallbackRemoteGetRequest;
+use Google\Web_Stories_Dependencies\AmpProject\RemoteRequest\FilesystemRemoteGetRequest;
+
+/**
+ * Optimization class.
+ *
+ * @since 1.1.0
+ */
+class Optimization {
+	/**
+	 * Optimizes a document.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param Document $document Document instance.
+	 *
+	 * @return void
+	 */
+	public function optimize_document( Document $document ) {
+		$errors = new ErrorCollection();
+		$this->get_optimizer()->optimizeDom( $document, $errors );
+
+		if ( count( $errors ) > 0 ) {
+			$error_messages = array_map(
+				static function( Error $error ) {
+					return ' - ' . $error->getCode() . ': ' . $error->getMessage();
+				},
+				iterator_to_array( $errors )
+			);
+			$document->head->appendChild(
+				$document->createComment( "\n" . __( 'AMP optimization could not be completed due to the following:', 'web-stories' ) . "\n" . implode( "\n", $error_messages ) . "\n" )
+			);
+		}
+	}
+
+	/**
+	 * Optimizer instance to use.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @return TransformationEngine Optimizer transformation engine to use.
+	 */
+	private function get_optimizer() {
+		$configuration = self::get_optimizer_configuration();
+
+		$fallback_remote_request_pipeline = new FallbackRemoteGetRequest(
+			new WpHttpRemoteGetRequest(),
+			new FilesystemRemoteGetRequest( LocalFallback::getMappings() )
+		);
+
+		$cached_remote_request = new CachedRemoteGetRequest( $fallback_remote_request_pipeline, WEEK_IN_SECONDS );
+
+		return new TransformationEngine(
+			$configuration,
+			$cached_remote_request
+		);
+	}
+
+	/**
+	 * Get the AmpProject\Optimizer configuration object to use.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @return Configuration Optimizer configuration to use.
+	 */
+	private static function get_optimizer_configuration() {
+		$transformers = Configuration::DEFAULT_TRANSFORMERS;
+
+		/**
+		 * Filter whether the AMP Optimizer should use server-side rendering or not.
+		 *
+		 * @since 1.1.0
+		 *
+		 * @param bool $enable_ssr Whether the AMP Optimizer should use server-side rendering or not.
+		 */
+		$enable_ssr = apply_filters( 'web_stories_enable_ssr', true );
+
+		// In debugging mode, we don't use server-side rendering, as it further obfuscates the HTML markup.
+		if ( ! $enable_ssr ) {
+			$transformers = array_diff(
+				$transformers,
+				[
+					AmpRuntimeCss::class,
+					ServerSideRendering::class,
+					TransformedIdentifier::class,
+				]
+			);
+		}
+
+		/**
+		 * Filter the configuration to be used for the AMP Optimizer.
+		 *
+		 * @since 1.1.0
+		 *
+		 * @param array $configuration Associative array of configuration data.
+		 */
+		$configuration = apply_filters(
+			'web_stories_amp_optimizer_config',
+			[
+				Configuration::KEY_TRANSFORMERS => $transformers,
+			]
+		);
```

Comment:
```suggestion
                 $configuration = [ Configuration::KEY_TRANSFORMERS => $transformers ];
		/**
		 * Filter the configuration to be used for the AMP Optimizer.
		 *
		 * @since 1.1.0
		 *
		 * @param array $configuration Associative array of configuration data.
		 */
		$configuration = apply_filters( 'web_stories_amp_optimizer_config', $configuration );
```